lib/puppet/provider/dns_record/dynect.rb

```diff
@@ -0,0 +1,149 @@
+      # Use strings over symbols because that's what dynect returns
+      # Makes compairison easier
+      when "A"
+          return {"address" => "#{resource[:content]}"}
+      when "CNAME"
+          # Append trailing period if needed
+          resource[:content] += "." if "#{resource[:content]}"[-1,1] != "."
+          return {"cname" => "#{resource[:content]}"}
+      when "TXT"
+          return {"txtdata" => "#{resource[:content]}"}
+      end
+  end
+
+  def logout
+      url = "https://api2.dynect.net/REST/Session/"
+      response = RestClient.delete(url,@headers)
+  end
+
+  def create
+    begin
+      Puppet.debug("Attempting to create record type #{resource[:type]} for #{resource[:name]} as #{resource[:content]}") if @action == "post"
+      Puppet.debug("Attempting to edit record type #{resource[:type]} for #{resource[:name]} as #{resource[:content]}") if @action == "put"
+      url = "https://api.dynect.net/REST/#{resource[:type]}Record/#{resource[:domain]}/#{resource[:name]}"
+      session_data = { :rdata => set_rdata, :ttl => resource[:ttl] }
+      response = RestClient.send(@action.to_sym, url,session_data.to_json, @headers)
+      obj = JSON.parse(response)
+      if obj['status'] == 'success'
+          # Publish the zone
+          url = "https://api2.dynect.net/REST/Zone/#{resource[:domain]}"
+          session_data = { "publish" => "true" }
+          response = RestClient.put(url,session_data.to_json,@headers)
+          obj = JSON.parse(response)
+          if obj['status'] == 'success'
+              Puppet.info("DynECT: Created #{resource[:type]} record for #{resource[:name]} with ttl #{resource[:ttl]}") if @action == "post"
+              if @old_ttl.nil? and @old_rdata
+                  Puppet.info("DynECT: Updated #{resource[:type]} record for #{resource[:name]} from #{@old_rdata.values[0]} to #{resource[:content]}")
+              elsif @old_ttl and @old_rdata.nil?
+                  Puppet.info("DynECT: Updated #{resource[:type]} record for #{resource[:name]} from ttl #{@old_ttl} to #{resource[:ttl]}")
+              else
+                  Puppet.info("DynECT: Updated #{resource[:type]} record for #{resource[:name]} from #{@old_rdata.values[0]} to #{resource[:content]} and ttl from #{@old_ttl} to #{resource[:ttl]}") if @action == "put"
+              end
+              logout
+          end
+      end
+    rescue Excon::Errors::UnprocessableEntity
+        Puppet.info("DynECT: #{e.response.body}")
+    end
+  end
+
+  def exists?
+      get_token
+      url = "https://api.dynect.net/REST/AllRecord/#{resource[:domain]}/"
+      response = RestClient.get(url,@headers)
+      obj = JSON.parse(response)
+      if obj['status'] != 'success'
+          #TODO error handle
+      end
+      # Find URL that contains the record ID, save it, and request more detailed information from dynect
+      index = obj['data'].index{|s| s.include?"#{resource[:type]}Record/#{resource[:domain]}/#{resource[:name]}"}
```

Comment:
Because of this you are pulling down the entire zone every time you want to validate a single item exists.  This is a sign that the provider should probably leverage `instances` and `prefetch` methods.

https://github.com/puppetlabs/puppetlabs-chronos/blob/master/lib/puppet/provider/chronos_job/chronos_job.rb#L36-L71